src/software/thunderscope/binary_context_managers/full_system.py

```diff
@@ -0,0 +1,182 @@
+import os
+import socket
+import logging
+import psutil
+import time
+import threading
+import google.protobuf.internal.encoder as encoder
+import google.protobuf.internal.decoder as decoder
+
+from subprocess import Popen
+from software.python_bindings import *
+from proto.import_all_protos import *
+from software.py_constants import *
+from extlibs.er_force_sim.src.protobuf.world_pb2 import (
+    SimulatorState,
+    SimBall,
+    SimRobot,
+)
+from software.thunderscope.binary_context_managers.util import *
+
+class FullSystem(object):
+
+    """ Full System Binary Context Manager """
+
+    def __init__(
+        self,
+        full_system_runtime_dir=None,
+        debug_full_system=False,
+        friendly_colour_yellow=False,
+        should_restart_on_crash=True,
+    ):
+        """Run FullSystem
+
+        :param full_system_runtime_dir: The directory to run the blue full_system in
+        :param debug_full_system: Whether to run the full_system in debug mode
+
+        """
+        self.full_system_runtime_dir = full_system_runtime_dir
+        self.debug_full_system = debug_full_system
+        self.friendly_colour_yellow = friendly_colour_yellow
+        self.full_system_proc = None
+        self.should_restart_on_crash = should_restart_on_crash
+
+        self.thread = threading.Thread(target=self.__restart__)
+
+    def __enter__(self):
+        """Enter the full_system context manager. 
+
+        If the debug mode is enabled then the binary is _not_ run and the
+        command to debug under gdb is printed. The  context manager will then
+        wait for the binary to be launched before continuing.
+
+        :return: full_system context managed instance
+
+        """
+        # Setup unix socket directory
+        try:
+            os.makedirs(self.full_system_runtime_dir)
+        except:
+            pass
+
+        self.full_system = "software/unix_full_system --runtime_dir={} {}".format(
+            self.full_system_runtime_dir,
+            "--friendly_colour_yellow" if self.friendly_colour_yellow else "",
+        )
+
+        if self.debug_full_system:
+
+            # We don't want to check the exact command because this binary could
+            # be debugged from clion or somewhere other than gdb
+            if not is_cmd_running(
+                [
+                    "unix_full_system",
+                    "--runtime_dir={}".format(self.full_system_runtime_dir),
+                ]
+            ):
+                logging.info(
+                    (
+                        f"""
+
+Debugging Fullsystem ==============
+1. Build the full system in debug mode:
+
+./tbots.py -d build unix_full_system
+
+2. Run the following binaries from src to debug full system:
+
+gdb --args bazel-bin/{full_system}
```

Comment:
I am not sure, but this variable seems to be undefined. 

Do mean mean `self` instead of `full_system`? Is this just a global variable that is accessed at run time?

Reply:
Yeah that seems incorrect. We don't need the `self.fullsystem` var defined above to be a member variable, so you can probably just remove `self` from that instead.